Libraries/Cordio/wsf/sources/targets/freertos/wsf_nvm.c

```diff
@@ -373,3 +432,65 @@ void WsfNvmEraseDataAll(WsfNvmCompEvent_t compCback)
     compCback(TRUE);
   }
 }
+
+/*************************************************************************************************/
+/*!
+ *  \brief  Defragment NVM
+ *  \param  copyBuf       Scratch buffer to temporarily copy NVM stored data.
+ *  \param  size          Size of copyBuf in bytes.
+ *  \return  TRUE if defragment successful. FALSE otherwise
+ * 
+ *  \note   Defragment should only be called when the storage is full and a record has been invalidated.
+ *          Furthermore, copyBuf must be at least the size of WSF NVM allocated flash.
+ */
+/*************************************************************************************************/
+bool_t WsfNvmDefragment(uint8_t *copyBuf, uint32_t size)
+{
+#if WSF_ASSERT_ENABLED == 1
+    WSF_ASSERT(copyBuf && size >= wsfNvmCb.totalSize);
+#else
+    if (!copyBuf || size < wsfNvmCb.totalSize) {
+        WSF_TRACE_INFO0("Not enough memory given to defragment NVM");
+        return FALSE;
+    }
+#endif
+
+    WsfCsEnter();
+
+    bool_t defragPossible = FALSE;
+    uint32_t currentOffset = WSF_NVM_START_ADDR;
+    uint32_t copyOffset = 0;
+    WsfNvmHeader_t header = { 0 };
+
+    /* Get the first header*/
+    PalFlashRead(&header, WSF_NVM_HEADER_SIZE, currentOffset);
+    while (currentOffset < wsfNvmCb.availAddr && header.id != WSF_NVM_UNUSED_FILECODE) {
+        const uint32_t fileSize = WSF_NVM_FILE_SIZE(header.len);
+
+        /*If the entry is valid copy it into the defragmentation buffer, if its reserved we can defrag it*/
+        if (header.id != WSF_NVM_RESERVED_FILECODE) {
+            PalFlashRead(&copyBuf[copyOffset], fileSize, currentOffset);
+            copyOffset += WSF_NVM_FILE_SIZE(header.len);
+        } else {
+            defragPossible = TRUE;
+        }
+
+        currentOffset += fileSize;
+        PalFlashRead(&header, WSF_NVM_HEADER_SIZE, currentOffset);
+    }
+
+    /*
+    If we defragged anything clear the sector and rewrite flash with defragmented data
+  */
+    if (defragPossible) {
+        WsfNvmEraseDataAll(NULL);
```

Comment:
I am seeing another HardFault inside this call, because inside `WsfNvmEraseDataAll`, `wsfNvmCb.availAddr` is not rounded up to the next nearest multiple of `wsfNvmCb.sectorSize`. So `PalFlashEraseSector` enters an endless loop of calling `MXC_FLC_PageErase`.

Reply:
Yes, i believe that is correct. I updated the macro to deal with alignment.

Reply:
I can get the hardfaults now, and I am working through them.

Reply:
I see that the first HardFault is fixed with the latest code that is pushed, but am still getting a hardfault here, if I try to defrag and `wsfNvmCb.availAddr` is not rounded up to the start of the next sector. This happens if you try to defrag before the NVM is full.

Reply:
It appears the baremetal implementation didn't have a bug fix. The code deletes sectors at a time, not bytes. The code just needed to be updated to not attempt to delete ```wsfNvm.availAddr``` worth of sectors